### DISCUSSION TITLE
Provide a magit-blame-quit-all option

### DISCUSSION THREAD

**Discussion by tomdl89** [magit/magit#4724] [Show and tell]
Often I will blame and then blame again, maybe a few times. Magit makes a new buffer for each revision, so when I am done, I call `magit-blame-quit` for every revision. I could also just `switch-to-buffer` to the initial revision and then call `magit-blame-quit` once, but that leaves all the other revisions in my buffer list, which I don't like. I propose new functionality called `magit-blame-quit-all` or somesuch, which kills buffers for all revisions except the initial one, and turns off magit-blame-mode in that. This would mean only having to invoke one command to get out of a recursive blame situation, which would be great.

**Comment by kyleam** [comment]
> only having to invoke one command to get out of a recursive blame situation

Yeah, I can see the appeal in that.  It'd take more time and thought before adding something like this to Magit itself, but fwiw you could try putting something like this in your config:

```elisp
(defun my/magit-blame-quit-all ()
  "Call `magit-blame-quit' in all blame buffers for current file.
Note that this links buffers through the file name, so it does
not operate across file renames."
  (interactive)
  (when-let ((fname (or magit-buffer-file-name
                        (buffer-file-name))))
    (dolist (buf (buffer-list))
      (setq buf (or (buffer-base-buffer buf) buf))
      (when (buffer-live-p buf)
        (with-current-buffer buf
          (when (and magit-blame-mode
                     (equal fname (or magit-buffer-file-name
                                      (buffer-file-name))))
            (magit-blame-quit)))))))

(define-key magit-blame-read-only-mode-map "Q" #'my/magit-blame-quit-all)
```

**Reply by tomdl89** [reply]
Thanks for the suggestion. I'll give it a try and feed back. I guess if this is implemented in magit the actual code would keep track of the "parent" and "children" to ensure we don't kill any non-related blaming buffers?

**Reply by tarsius** [reply]
Here's a slightly updated version:

```elisp
(defun magit-blame-quit-all ()
  "Turn off Magit-Blame mode in all buffers visiting versions of current file.
Buffers visiting different versions of the same file are determined by
comparing file names, so this does not work across renames."
  (interactive)
  (if-let ((name (magit-buffer-file-name)))
      (dolist (buf (buffer-list))
        (setq buf (or (buffer-base-buffer buf) buf))
        (when (and (buffer-live-p buf)
                   (buffer-local-value 'magit-blame-mode buf))
          (with-current-buffer buf
            (when (equal (magit-buffer-file-name) name)
              (magit-blame-mode -1)))))
    (user-error "Not visiting a file or blob")))
```

I have decided against adding something like this, but I am adding an entry to my private todo list to maybe eventually implement "cross-version navigation recording and eventual cleanup". Beside blaming, `magit-{previous,next}-blob` also fall into that category.

(This should also kill buffer, but only if they weren't open before. Unkilled buffers should only be removed from window's buffer history. Actually maybe buffers should not be killed at all. We cache blob buffers for performance reasons (since we added diff fontification. So this interacts with other forms of blob cache invalidation. It's complicated.)

The reason I am not adding the above command is that it just doesn't cut it for me. Either we cleanup "the whole session" or let it be.
